src/prof_sys.c

```diff
@@ -263,24 +263,28 @@ prof_backtrace_impl(prof_bt_t *bt) {
 }
 #else
 static void
-prof_backtrace_impl(prof_bt_t *bt) {
+prof_backtrace_impl(void **vec, unsigned *len, unsigned max_len) {
 	cassert(config_prof);
 	not_reached();
 }
 #endif
 
-
-void (* JET_MUTABLE prof_backtrace_hook)(prof_bt_t *bt) = &prof_backtrace_impl;
-
 void
 prof_backtrace(tsd_t *tsd, prof_bt_t *bt) {
 	cassert(config_prof);
 	pre_reentrancy(tsd, NULL);
-	prof_backtrace_hook(bt);
+	prof_backtrace_hook_t prof_backtrace_hook = prof_backtrace_hook_get();
+	prof_backtrace_hook(bt->vec, &bt->len, PROF_BT_MAX);
 	post_reentrancy(tsd);
 }
 
-void prof_unwind_init() {
+void
+prof_hooks_init() {
+	prof_backtrace_hook_set(&prof_backtrace_impl);
+}
+
+void
+prof_unwind_init() {
```

Comment:
Merge the two inits? Whenever you have two relevant inits, the order may complicate things. Merging into one will avoid such concern.

Reply:
What would you call the function that initializes both hooks and `unwind`? To me, they are different, but the order is irrelevant.

Reply:
I was considering the backtrace hook as a higher level feature of the unwind part, in which case calling the backtrace_hook_init within the unwind_init would be fine. But just noticed that the two hooks init are grouped together which looks good too.

In this case I'd say invoking the hooks_init last. Although it doesn't matter for the current impl, it's logically higher level (it's not that crazy if some hook init assumes the underlying unwinder is functional).

Reply:
Addressed